In [2]:
import numpy as np
import pandas as pd

# Qustion 0

In [81]:
sample_list = [(1, 3, 5), (0, 1, 2), (1, 9, 8)]
op = []
for m in range(len(sample_list)):
    li = [sample_list[m]]
        for n in range(len(sample_list)):
            if (sample_list[m][0] == sample_list[n][0] and
                    sample_list[m][3] != sample_list[n][3]):
                li.append(sample_list[n])
        op.append(sorted(li, key=lambda dd: dd[3], reverse=True)[0])
res = list(set(op))

IndentationError: unexpected indent (Temp/ipykernel_6856/1593771623.py, line 5)

- (a) What does the code do?

It first classifies tuples in the list according to the value of the first item, then select the tuple with the largest last item in each category, and returns all these tuples without repitition.
- (b) Code review
   - Working part
   
     The code can express the writer's thoughts, but unfortunately, it doesn't work. There are several syntax errors in this code. The second "for" loop is supposed to be aligned with the assignment line, and the list range is zero to length-1, not 0 to length.
   - Style
     
     It doesn't follow the style guidelines.
   - Structure
     
     It is clearly structured. The writer used two loops and it's easy to see his idea.
   - Efficiency
   
     The efficiency can be improved. For example, this code uses `sample_list[m]` several times. for the "if" statement in the second loop, it can be replaced by `li[0]`.And there's no need to judge whether `sample_list[m][3]` is equal to `sample_list[n][3]`, because the `set` part will remove duplicate values.
     
   - Suggestion
     First, maybe the code writer should focus more on the use of "tab", and check the alignment and foot numbers in the code more carefully. Second, when a variable is used many times, maybe the writer can notice it and temporarily store it in a parameter. Third, focus more on the details of the code. Moreover, the writer may use a function and do your work in a more general way.    

# Question 1

In [82]:
def f(n, k = 4, low = 0, high = 10):
    '''
    return a list of n-tuples with k random integers between low and high.
    Parameters
    ----------
    n : int TYPE
        The number of tuples.
    k : int TYPE, optional
        The number of random integers. The default is 4.
    low :int TYPE, optional
        The lower bound. The default is 0.
    high :int TYPE, optional
        The upper bound. The default is 10.

    Returns
    -------
    list0 :list TYPE
        A list of n-tuples with k random integers between low and high.

    '''
    list0 = []
    for i in range(n):
        list0.append(tuple(np.random.randint(low,high,size=k)))
    #assert test
    #list0 = "test"
    #list0 = [2,3]
    assert type(list0)==list, "Return should be a list of tuples."
    for t in list0:
        assert type(t)== tuple,"Return should be a list of tuples."
    return list0
print(f(5))

[(6, 5, 0, 0), (8, 6, 3, 4), (5, 9, 0, 5), (2, 9, 6, 4), (5, 5, 5, 7)]


# Question 2

In [83]:
def given_max1(list1, item1 = 0, item2 = 3):
    '''
    Classify the tuples in list1 through item1, and choose the max of item2.
    Parameters
    ----------
    list1 : list TYPE
        The list generated by the function in a.
    item1 : int TYPE, optional
        The classify column number. The default is 0.
    item2 : TYPE, optional
        The sorted column number. The default is 3.

    Returns
    -------
    list TYPE
        a list of these "max" tuples.

    '''
    sample_list = list1
    op = []
    for m in range(len(sample_list)):
        li = [sample_list[m]]
        for n in range(len(sample_list)):
            if (sample_list[m][item1] == sample_list[n][item1] and
                    sample_list[m][item2] != sample_list[n][item2]):
                li.append(sample_list[n])
        op.append(sorted(li, key=lambda dd: dd[item2], reverse=True)[0])
    return list(set(op))
print(given_max1([(1, 3, 5, 6), (0, 1, 2, 3), (1, 9, 8, 10),(1, 3, 5, 10)]))

[(1, 9, 8, 10), (0, 1, 2, 3), (1, 3, 5, 10)]


In [30]:
def given_max2(list1, item1 = 0, item2 = 3):
    '''
    Classify the tuples in list1 through item1, and choose the max of item2.
    Parameters
    ----------
    list1 : list TYPE
        The list generated by the function in a.
    item1 : int TYPE, optional
        The classify column number. The default is 0.
    item2 : TYPE, optional
        The sorted column number. The default is 3.

    Returns
    -------
    list TYPE
        a list of these "max" tuples.

    '''
    op, leng = [], len(list1)
    for m in range(leng):
        a = list1[m]
        li = [a]
        for n in range(leng):
            if a[item1] == list1[n][item1]:
                li.append(list1[n])
        op.append(sorted(li, key=lambda dd: dd[item2], reverse=True)[0])
    return list(set(op))
print(given_max2([(1, 3, 5, 6), (0, 1, 2, 3), (1, 9, 8, 10),(1, 3, 5, 10)]))

[(1, 9, 8, 10), (0, 1, 2, 3), (1, 3, 5, 10)]


In [84]:
def given_max3(list1, item1 = 0, item2 = 3):
    '''
    Classify the tuples in list1 through item1, and choose the max of item2.
    Parameters
    ----------
    list1 : list TYPE
        The list generated by the function in a.
    item1 : int TYPE, optional
        The classify column number. The default is 0.
    item2 : TYPE, optional
        The sorted column number. The default is 3.

    Returns
    -------
    list TYPE
        a list of these "max" tuples.

    '''
    dict1 = {}
    for tup in list1:
        a = tup[item1]
        if dict1.get(a) != None:
            if dict1[a][item2] < tup[item2]:
                dict1[a] = tup
        else:
            dict1[a] = tup
    return list(dict1.values())
print(given_max3([(1, 3, 5, 6), (0, 1, 2, 3), (1, 9, 8, 10),(1, 3, 5, 10)]))

[(1, 9, 8, 10), (0, 1, 2, 3)]


# Question3

a)

In [3]:
df1718 = pd.read_sas('DEMO_2017-2018.XPT', format = 'xport')
df1516 = pd.read_sas('DEMO_2015-2016.XPT', format = 'xport')
df1314 = pd.read_sas('DEMO_2013-2014.XPT', format = 'xport')
df1112 = pd.read_sas('DEMO_2011-2012.XPT', format = 'xport')
interest = ["SEQN", "RIDAGEYR", "RIDRETH3", "DMDEDUC2", "DMDMARTL", "RIDSTATR", "SDMVPSU", "SDMVSTRA", "WTMEC2YR", "WTINT2YR"]
z = []
y4 = df1718[interest]
m, n = y4.shape[0],y4.shape[1]
y4.insert(n,"YEAROFDATA",["17-18"for i in range(m)])
z.append(y4)
y3 = df1516[interest]
m, n = y3.shape[0],y3.shape[1]
y3.insert(n,"YEAROFDATA",["15-16"for i in range(m)])
z.append(y3)
y2 = df1314[interest]
m, n = y2.shape[0],y2.shape[1]
y2.insert(n,"YEAROFDATA",["13-14"for i in range(m)])
z.append(y2)
y1 = df1112[interest]
m, n = y1.shape[0],y1.shape[1]
y1.insert(n,"YEAROFDATA",["11-12"for i in range(m)])
z.append(y1)
res = pd.concat(z, axis = 0, ignore_index = True)
res.columns = [x.lower() for x in interest]+["yearofdata"]
res = pd.DataFrame.convert_dtypes(res)
print(res)

        seqn  ridageyr  ridreth3  dmdeduc2  dmdmartl  ridstatr  sdmvpsu  \
0      93703       2.0         6      <NA>      <NA>         2        2   
1      93704       2.0         3      <NA>      <NA>         2        1   
2      93705      66.0         4         2         3         2        2   
3      93706      18.0         6      <NA>      <NA>         2        2   
4      93707      13.0         7      <NA>      <NA>         2        1   
...      ...       ...       ...       ...       ...       ...      ...   
39151  71912      40.0         3         1         1         2        1   
39152  71913      18.0         6      <NA>      <NA>         2        1   
39153  71914      10.0         3      <NA>      <NA>         2        2   
39154  71915      60.0         3         5         5         2        3   
39155  71916      16.0         3      <NA>      <NA>         2        1   

       sdmvstra      wtmec2yr      wtint2yr yearofdata  
0           145   8539.731348   9246.49186

b)the round-trip part ( to make sure it doesn't miss any information.)

In [86]:
pd.DataFrame.to_pickle(res, "nhane.pkl")
pd.DataFrame.to_feather(res, "nhane.feather")
res_pi = pd.read_pickle("nhane.pkl")
print(res_pi)
res_f = pd.read_feather("nhane.feather")
print(res_f)

        seqn  ridageyr  ridreth3  dmdeduc2  dmdmartl  ridstatr  sdmvpsu  \
0      93703       2.0         6      <NA>      <NA>         2        2   
1      93704       2.0         3      <NA>      <NA>         2        1   
2      93705      66.0         4         2         3         2        2   
3      93706      18.0         6      <NA>      <NA>         2        2   
4      93707      13.0         7      <NA>      <NA>         2        1   
...      ...       ...       ...       ...       ...       ...      ...   
39151  71912      40.0         3         1         1         2        1   
39152  71913      18.0         6      <NA>      <NA>         2        1   
39153  71914      10.0         3      <NA>      <NA>         2        2   
39154  71915      60.0         3         5         5         2        3   
39155  71916      16.0         3      <NA>      <NA>         2        1   

       sdmvstra      wtmec2yr      wtint2yr yearofdata  
0           145   8539.731348   9246.49186

We can see the data after round-trip format is the same.

In [4]:
dff1718 = pd.read_sas('OHXDEN_2017-2018.XPT', format = 'xport')
dff1516 = pd.read_sas('OHXDEN_2015-2016.XPT', format = 'xport')
dff1314 = pd.read_sas('OHXDEN_2013-2014.XPT', format = 'xport')
dff1112 = pd.read_sas('OHXDEN_2011-2012.XPT', format = 'xport')
total = dff1718.columns
interest1 = ['SEQN', 'OHDDESTS'] + [x for x in total if (x[0:3] == "OHX" and (x[5:7] == "TC" or x[5:8] == "CTC")) ]
for x in interest1:
    if x not in dff1516.columns:
        interest1.remove(x)
for x in interest1:
    if x not in dff1314.columns:
        interest1.remove(x)
for x in interest1:
    if x not in dff1112.columns:
        interest1.remove(x)
z1 = []
y8 = dff1718[interest1]
m, n = y8.shape[0],y8.shape[1]
y8.insert(n,"YEAROFDATA",["17-18"for i in range(m)])
z1.append(y8)
y7 = dff1516[interest1]
m, n = y7.shape[0],y7.shape[1]
y7.insert(n,"YEAROFDATA",["15-16"for i in range(m)])
z1.append(y7)
y6 = dff1314[interest1]
m, n = y6.shape[0],y6.shape[1]
y6.insert(n,"YEAROFDATA",["13-14"for i in range(m)])
z1.append(y6)
y5 = dff1112[interest1]
m, n = y5.shape[0],y5.shape[1]
y5.insert(n,"YEAROFDATA",["11-12"for i in range(m)])
z1.append(y5)
res1 = pd.concat(z1, axis = 0, ignore_index = True)
res1.columns = [x.lower() for x in interest1]+["yearofdata"]
res1 = pd.DataFrame.convert_dtypes(res1)
print(res1)
print(res1['seqn'])

D:\anaconda\lib\site-packages\pandas\io\sas\sas_xport.py:475: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  df[x] = v


        seqn  ohddests  ohx01tc  ohx02tc  ohx03tc  ohx04tc  ohx05tc  ohx06tc  \
0      93703         1        4        4        4        1        1        1   
1      93704         1        4        4        4        1        1        1   
2      93705         1        4        2        4        2        2        2   
3      93706         1        4        2        2        2        2        2   
4      93707         1        4        4        2        1        1        1   
...      ...       ...      ...      ...      ...      ...      ...      ...   
35904  71912         1        4        2        2        2        2        2   
35905  71913         1        2        2        2        2        2        2   
35906  71914         1        4        4        2        1        1        1   
35907  71915         1        2        2        2        2        2        2   
35908  71916         1        4        2        2        2        2        2   

       ohx07tc  ohx08tc  ...  ohx23ctc 

In [79]:
pd.DataFrame.to_pickle(res1, "OH.pkl")
pd.DataFrame.to_feather(res1, "OH.feather")
res1_pi = pd.read_pickle("OH.pkl")
print(res1_pi)
res1_f = pd.read_feather("OH.feather")
print(res1_f)

        seqn  ohddests  ohx01tc  ohx02tc  ohx03tc  ohx04tc  ohx05tc  ohx06tc  \
0      93703         1        4        4        4        1        1        1   
1      93704         1        4        4        4        1        1        1   
2      93705         1        4        2        4        2        2        2   
3      93706         1        4        2        2        2        2        2   
4      93707         1        4        4        2        1        1        1   
...      ...       ...      ...      ...      ...      ...      ...      ...   
35904  71912         1        4        2        2        2        2        2   
35905  71913         1        2        2        2        2        2        2   
35906  71914         1        4        4        2        1        1        1   
35907  71915         1        2        2        2        2        2        2   
35908  71916         1        4        2        2        2        2        2   

       ohx07tc  ohx08tc  ...  ohx23ctc 

In [6]:
from os.path import exists
# file location: -------------------------------------------------------------
path = './'


# column maps: ---------------------------------------------------------------
# new names for demo cols
demo_cols = {
    'SEQN': 'id',
    'RIDAGEYR': 'age',
    'RIAGENDR': 'gender',
    'RIDRETH3': 'race',
    'DMDEDUC2': 'education',
    'DMDMARTL': 'marital_status',
    'RIDSTATR': 'exam_status',
    'SDMVPSU': 'psu',
    'SDMVSTRA': 'strata',
    'WTMEC2YR': 'exam_wt',
    'WTINT2YR': 'interview_wt'
    }



# new names for ohx cols
ohx_cols = {'SEQN': 'id', 'OHDDESTS': 'dentition_status'}
tc_cols = {'OHX' + str(i).zfill(2) + 'TC':
           'tc_' + str(i).zfill(2) for i in range(1, 33)}
ctc_cols = {'OHX' + str(i).zfill(2) + 'CTC':
            'ctc_' + str(i).zfill(2) for i in range(2, 32)}
_, _ = ctc_cols.pop('OHX16CTC'), ctc_cols.pop('OHX17CTC')

ohx_cols.update(tc_cols)
ohx_cols.update(ctc_cols)

# columns to convert to integer
demo_int = ('id', 'age', 'psu', 'strata')
ohx_int = ('id', )
# levels for categorical variables
demo_cat = {
    'gender': {1: 'Male', 2: 'Female'},
    'race': {1: 'Mexican American',
             2: 'Other Hispanic',
             3: 'Non-Hispanic White',
             4: 'Non-Hispanic Black',
             6: 'Non-Hispanic Asian',
             7: 'Other/Multiracial'
             },
    'education': {1: 'Less than 9th grade',
                  2: '9-11th grade (Includes 12th grade with no diploma)',
                  3: 'High school graduate/GED or equivalent',
                  4: 'Some college or AA degree',
                  5: 'College graduate or above',
                  7: 'Refused',
                  9: "Don't know"
                  },
    'marital_status': {1: 'Married',
                       2: 'Widowed',
                       3: 'Divorced',
                       4: 'Separated',
                       5: 'Never married',
                       6: 'Living with partner',
                       77: 'Refused',
                       99: "Don't know"
                       },
    'exam_status': {1: 'Interviewed only',
                    2: 'Both interviewed and MEC examined'
                    }
    }

ohx_cat = {
    'dentition_status': {1: 'Complete', 2: 'Partial', 3: 'Not Done'}
    }

tc = {
      1: 'Primary tooth present',
      2: 'Permanent tooth present',
      3: 'Dental Implant',
      4: 'Tooth not present',
      5: 'Permanent dental root fragment present',
      9: 'Could not assess'
      }
ctc = (
 {
  'A': 'Primary tooth with a restored surface condition',
  'D': 'Sound primary tooth',
  'E': 'Missing due to dental disease',
  'F': 'Permanent tooth with a restored surface condition',
  'J':
    'Permanent root tip is present but no restorative replacement is present',
  'K': 'Primary tooth with a dental carious surface condition',
  'M': 'Missing due to other causes',
  'P':
    'Missing due to dental disease but replaced by a removable restoration',
  'Q':
    'Missing due to other causes but replaced by a removable restoration',
  'R':
    'Missing due to dental disease but replaced by a fixed restoration',
  'S': 'Sound permanent tooth',
  'T':
    'Permanent root tip is present but a restorative replacement is present',
  'U': 'Unerupted',
  'X': 'Missing due to other causes but replaced by a fixed restoration',
  'Y': 'Tooth present, condition cannot be assessed',
  'Z': 'Permanent tooth with a dental carious surface condition'
 })
# read data: -----------------------------------------------------------------
base_url = 'https://wwwn.cdc.gov/Nchs/Nhanes/'
cohorts = (
    ('2011-2012', 'G'),
    ('2013-2014', 'H'),
    ('2015-2016', 'I'),
    ('2017-2018', 'J')
    )
# demographic data
demo_file = path + '/demo.feather'
if exists(demo_file):
    demo = pd.read_feather(demo_file)
else:
    demo_cohorts = {}
    for cohort, label in cohorts:

        # read data and subset columns
        url = base_url + cohort + '/DEMO_' + label + '.XPT'
        dat = pd.read_sas(url).copy()
        dat = dat[list(demo_cols.keys())].rename(columns=demo_cols)

        # assign cohort and collect
        dat['cohort'] = cohort
        demo_cohorts.update({cohort: dat})

    # concatenate and save
    demo = pd.concat(demo_cohorts, ignore_index=True)
 
    # categorical variables
    for col, d in demo_cat.items():
        demo[col] = pd.Categorical(demo[col].replace(d))
    demo['cohort'] = pd.Categorical(demo['cohort'])

    # integer variables
    for col in demo_int:
        demo[col] = pd.to_numeric(demo[col], downcast='integer')

    demo.to_feather(demo_file)
demo.shape
# dentition data
ohx_file = path + '/ohx.feather'
if exists(ohx_file):
    ohx = pd.read_feather(ohx_file)
else:
    ohx_cohorts = {}
    for cohort, label in cohorts:

        # read data and subset columns
        url = base_url + cohort + '/OHXDEN_' + label + '.XPT'
        dat = pd.read_sas(url).copy()
        dat = dat[list(ohx_cols.keys())].rename(columns=ohx_cols)

        # assign cohort and collect
        dat['cohort'] = cohort
        ohx_cohorts.update({cohort: dat})
 
    # concatenate
    ohx = pd.concat(ohx_cohorts, ignore_index=True)

    # categorical variables
    for col, d in ohx_cat.items():
        ohx[col] = pd.Categorical(ohx[col].replace(d))
    
    for col in tc_cols.values():
        ohx[col] = pd.Categorical(ohx[col].replace(tc))

    # ctc columns get read in as bytes
    for col in ctc_cols.values():
        ohx[col] = ohx[col].apply(lambda x: x.decode('utf-8'))
        ohx[col] = pd.Categorical(ohx[col].replace(ctc))

    ohx['cohort'] = pd.Categorical(ohx['cohort'])
    # integer variables
    for col in ohx_int:
        ohx[col] = pd.to_numeric(ohx[col], downcast='integer')

    # save
    ohx.to_feather(ohx_file)
ohx.shape
# ---

D:\anaconda\lib\site-packages\pandas\io\sas\sas_xport.py:475: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  df[x] = v


(35909, 63)

In [8]:
print(ohx)
print(demo)

           id dentition_status                    tc_01  \
0       62161         Complete        Tooth not present   
1       62162         Complete        Tooth not present   
2       62163         Complete        Tooth not present   
3       62164         Complete        Tooth not present   
4       62165         Complete        Tooth not present   
...       ...              ...                      ...   
35904  102952         Complete  Permanent tooth present   
35905  102953         Complete  Permanent tooth present   
35906  102954         Complete  Permanent tooth present   
35907  102955         Complete        Tooth not present   
35908  102956         Complete        Tooth not present   

                         tc_02                    tc_03  \
0      Permanent tooth present  Permanent tooth present   
1            Tooth not present        Tooth not present   
2      Permanent tooth present  Permanent tooth present   
3      Permanent tooth present  Permanent tooth present

c)For question a, we have 39156 cases. In b, there are 35909 cases.